### DISCUSSION TITLE
set_writable(), but only for value

### DISCUSSION THREAD

**Discussion by tomasgiden** [FreeOpcUa/opcua-asyncio#450] [Q&A]
If I set a variable to be writable with set_writable() and a buggy client then writes to it but with another type than the variable was created with, the server happily accepts the write, changing both the type and the value in the process. I tried it from boolean to uint16_t. If this is not a bug, then how to I make the value writable but not any other information related to the node?

**Comment by swamper123** [comment]
I can confirm that issue.

Minimal example to reproduce:

Client:

```python
import asyncio
import sys
# sys.path.insert(0, "..")
import logging
from asyncua import Client, Node, ua

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger('asyncua')

async def main():
    url = 'opc.tcp://0.0.0.0:4840/freeopcua/server/'
    async with Client(url=url) as client:
        uri = 'http://examples.freeopcua.github.io'
        idx = await client.get_namespace_index(uri)

        var = await client.nodes.root.get_child(["0:Objects", f"{idx}:MyObject", f"{idx}:Problematic_Bool"])
        print("Problematic_Bool", var, await var.read_value())

        await var.write_value(123, ua.VariantType.Int32)
        print("Problematic_Bool", var, await var.read_value())


if __name__ == '__main__':
    asyncio.run(main())

```


Server:

```python
import logging
import asyncio
import sys
sys.path.insert(0, "..")

from asyncua import ua, Server
from asyncua.common.methods import uamethod

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger('asyncua')

async def main():
    server = Server()
    await server.init()
    server.set_endpoint('opc.tcp://0.0.0.0:4840/freeopcua/server/')

    uri = 'http://examples.freeopcua.github.io'
    idx = await server.register_namespace(uri)

    myobj = await server.nodes.objects.add_object(idx, 'MyObject')
    myvar = await myobj.add_variable(idx, 'Problematic_Bool', ua.Variant(True, ua.VariantType.Boolean))
    await myvar.set_writable()

    _logger.info('Starting server!')
    async with server:
        while True:
            await asyncio.sleep(1)

if __name__ == '__main__':
    asyncio.run(main())
```

That leads in UAExpert to:

![grafik](https://user-images.githubusercontent.com/24408657/108062717-7e1b0680-705a-11eb-9ccc-6c8f5d234a56.png)

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.10.4/

A Server shall return a Bad_TypeMismatch error if the data type of the written value is not the same type or subtype of the Attribute’s DataType.

**Reply by swamper123** [reply]
I tried to exploit that behaviour with UaExpert, but wasn't successfull (I think they got a mechanism in their code, which doesn't allow such an invalid request).

Anyway, it think it should return a _Bad_TypeMismatch_.

I guess there is a check missing, if the written value is the same Type as the existing Node, around here:

https://github.com/FreeOpcUa/opcua-asyncio/blob/659260c371c84c66985f58368047539f039b4a00/asyncua/server/address_space.py#L56

**Reply by AndreasHeine** [reply]
subtypechecking !!!

**Reply by oroulet** [reply]
yes we do not do any type check in python server. neither in client. If you implement something server size, make sure it is not too expensive.

**Reply by oroulet** [reply]
and subtypechecking s going to be a nightmare. I am supposing many servers do not do it. But you can check the VariantType at least. That one should be cheap

**Reply by tomasgiden** [reply]
I have also noticed that a server can, by writing to a previously created variable, change type of that value. At least it is an server internal interface so not as bad as through OPC UA, but anyways, would be good to get an error message when one makes such a typo.

**Reply by oroulet** [reply]
we also allow to write any variable arttribute. It might make sense to make attrtibuted not writtable by default unless you are admin. Not sure there is code for that but it should not be so hard

**Reply by swamper123** [reply]
@oroulet 
> 
> and subtypechecking s going to be a nightmare. I am supposing many servers do not do it. But you can check the VariantType at least. That one should be cheap

in address_space.py:

![grafik](https://user-images.githubusercontent.com/24408657/108081436-42d80200-7071-11eb-9ff0-8eb123045452.png)
Green -  added Check
Red - VariantType

I thought that would be a quick one, but it might be a bit more complicated. Like here the Node "Namespaces" i=2255 shall be written with a value. The Node exists with `VariantType.Null ` and the value, which shall be written, is from type `VariantType.String` (which would be correct). This is caused while internal_server is in init.

Either we have to workaround that thing, when the server is initialising, or find another, better place to check.

**Reply by oroulet** [reply]
this looks like the correct place to implement the check. That the node over as the VariantType.Null sounds like a bug. There is already a method in InternalServer called fix_address_space or similar you could fix that one if this is the only one. 
Otherwise maybe we have some wrong default when creating address space? What is that node?

**Reply by swamper123** [reply]
I cought ~26 Nodes having this issue. I see what I can do, if I got some time, this week is pretty short for me.

**Reply by swamper123** [reply]
Well, it was a while ago, but had a look on it again. These 26 Node values, which have VariantType.Null, but should have another one, get this wrong Type, while standard_Address_space_services.py is running.

These 26 Nodes are from ServerStatus (and below), OperationLimits (and below), NamespaceArray, ServerArray and ServiceLevels. They got the Vatiant.Null Type first, but in the later building process, they get their real Type. I made an ignore list with these Nodes.

There are still issues, in case of broken Types, but I am on it.
